### PR TITLE
[ISHINTW-1576] Sync Diff Cpks From S3

### PR DESCRIPTION
**Pull request by RongRongTeng**
Cli command 
`aily design diff_download <base_version>  <target_version> <asset_type> <bucket>  [--client_dir client-repo-directory] [--design_dir design-repo-directory] [--skip_update] [-p path|--path path]`

Example:
`aily design diff_download develop master tutorial0-ios ishin-global-gogeta`
螢幕快照 2019-05-27 上午10 20 18

### CODE REVIEW

File: lib/aily/cli.rb [hunk 1]
```diff
@@ -33,6 +33,26 @@
...
+    diff.arg_name 'path'
+    diff.flag [:p, :path]
+
+    diff.switch [:skip_update]
+
+    diff.action do |global_options, options, args|
+      Aily::Design::Differ.download(args, options)
+    end
+  end
+end
```
Comment (cloud1985xp-aktsk) [inline lib/aily/cli.rb]: There already had a block of design command at line: 106

File: lib/aily/design/differ.rb [hunk 2]
```diff
@@ -7,15 +8,46 @@ class Differ
...
+
+          service = Aily::Utils::S3Service.new(bucket, rule['s3_path'], nil, client_directory, files: differ.diff_cpks)
+          logger.info "Start to download changed cpk files from(#{bucket}) to client repo(#{options[:client_dir]})"
+          jobs = service.execute(service.generate_cpk_jobs_list)
+
+          logger.info "Done!"
+        rescue => e
+          puts e.message
+          puts e.backtrace
+          raise e
```
Comment (cloud1985xp-aktsk) [inline lib/aily/design/differ.rb]: For debugging?

File: lib/aily/design/differ.rb [hunk 3]
```diff
@@ -7,15 +8,46 @@ class Differ
...
+          logger.info "Comparing changed files between #{base} and #{target} of #{type}."
+          differ = new(base, target, design_directory, type,
+                options.slice(:skip_update, :path)
+              )
+          logger.info "#{differ.reports}"
+          logger.info "Changed cpks files: #{differ.diff_cpks}"
+
+          service = Aily::Utils::S3Service.new(bucket, rule['s3_path'], nil, client_directory, files: differ.diff_cpks)
+          logger.info "Start to download changed cpk files from(#{bucket}) to client repo(#{options[:client_dir]})"
+          jobs = service.execute(service.generate_cpk_jobs_list)
```
Comment (cloud1985xp-aktsk) [inline lib/aily/design/differ.rb]: what does jobs mean for?  seems it always returns true?

File: lib/aily/design/differ.rb (hunk 2 above)
Comment (RongRongTeng) [reply lib/aily/design/differ.rb]: Copy from `master/differ.rb` 🤣

**Comment by HannahLo** [comment]
It will be great if add CLI command to README file.

File: lib/aily/utils/s3_service.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+      def execute(jobs)
+        parallel_execute(jobs)
+      end
+
+      def generate_cpk_jobs_list
+        @files.flat_map do |file|
+          objects = @bucket.objects(prefix: "#{@path}/#{@timestamp}/cpk/#{file}").collect(&:key)
+          objects.map do |object|
+          to_path = object.gsub("#{@path}/#{@timestamp}/cpk", @directory)
```
Comment (cloud1985xp-aktsk) [inline lib/aily/utils/s3_service.rb]: indent?

File: lib/aily/utils/s3_service.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            begin
+              while cmd = work_q.pop(true)
+                puts "Executing: #{cmd.ljust(100)} Item: #{commands.index(cmd)+1}/#{commands.size}\n"
+                eval cmd
+              end
+            rescue ThreadError
+            end
+          end
+        end
+        workers.map(&:join)
```
Comment (cloud1985xp-aktsk) [inline lib/aily/utils/s3_service.rb]: .each ?

File: lib/aily/utils/s3_service.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      end
+
+      def parallel_execute(commands)
+        work_q = Queue.new
+        commands.to_a.each { |x| work_q.push x }
+        workers = (0...4).map do
+          Thread.new do
+            begin
+              while cmd = work_q.pop(true)
+                puts "Executing: #{cmd.ljust(100)} Item: #{commands.index(cmd)+1}/#{commands.size}\n"
```
Comment (cloud1985xp-aktsk) [inline lib/aily/utils/s3_service.rb]: > commands.size

extract to out of the loop

File: lib/aily/utils/s3_service.rb [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                eval cmd
+              end
+            rescue ThreadError
```
Comment (cloud1985xp-aktsk) [inline lib/aily/utils/s3_service.rb]: need to handle it?

Comment (RongRongTeng) [reply lib/aily/utils/s3_service.rb]: Just want to rescue queue empty error

**Comment by cloud1985xp-aktsk** [comment]
👀

**Pull request merged**